test/gtest/ucp/test_ucp_proto_mock.cc

```diff
@@ -0,0 +1,343 @@
+    using iface_attr_func_t = void (*)(uct_iface_attr&);
+
+    mock_iface()
+    {
+        ucs_assert(m_self == nullptr);
+        m_self = this;
+    }
+
+    ~mock_iface()
+    {
+        cleanup();
+        m_self = nullptr;
+    }
+
+    void cleanup()
+    {
+        m_mock.cleanup();
+    }
+
+    void set_mock_iface_attr(const std::string &tl_name, iface_attr_func_t cb)
+    {
+        mock_tl(tl_name);
+        m_iface_attrs_funcs[tl_name] = cb;
+    }
+
+    void skip_on_error()
+    {
+        if (!m_error.empty()) {
+            UCS_TEST_SKIP_R(m_error);
+        }
+    }
+
+private:
+    void mock_tl(const std::string &tl_name)
+    {
+        uct_component_h component;
+        ucs_list_for_each(component, &uct_components_list, list) {
+            uct_tl_t *tl;
+            ucs_list_for_each(tl, &component->tl_list, list) {
+                if (tl_name == tl->name) {
+                    EXPECT_EQ(0, m_tls.count(tl_name));
+                    m_tls[tl_name] = tl;
+                    m_mock.setup(&tl->iface_open, iface_open_mock);
+                }
+            }
+        }
+    }
+
+    static ucs_status_t
+    iface_open_mock(uct_md_h md, uct_worker_h worker,
+                    const uct_iface_params_t *params,
+                    const uct_iface_config_t *config, uct_iface_h *iface_p)
+    {
+        uct_tl_t *tl = m_self->m_tls[params->mode.device.tl_name];
+        ucs_status_t status;
+
+        status = m_self->m_mock.orig_func(&tl->iface_open, md, worker, params,
+                                          config, iface_p);
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        uct_base_iface_t *base      = ucs_derived_of(*iface_p, uct_base_iface_t);
+        m_self->m_iface_names[base] = params->mode.device.tl_name;
+        m_self->m_mock.setup(&(*iface_p)->ops.iface_query, iface_query_mock);
+        return status;
+    }
+
+    static ucs_status_t
+    iface_query_mock(uct_iface_h iface, uct_iface_attr_t *iface_attr)
+    {
+        ucs_status_t status = m_self->m_mock.orig_func(
+                                    &iface->ops.iface_query, iface, iface_attr);
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        uct_base_iface_t *base  = ucs_derived_of(iface, uct_base_iface_t);
+        std::string &iface_name = m_self->m_iface_names[base];
+        auto it                 = m_self->m_iface_attrs_funcs.find(iface_name);
+        unsigned max_dev_path   = iface_attr->dev_num_paths;
+        (it->second)(*iface_attr);
+
+        /*
+         * It's hard to mock more paths than are actually supported by the
+         * hardware, so in this case we want to skip the test. Returning an
+         * error or skipping the test directly from this point is problematic,
+         * therefore we remember the error and check it in skip_on_error() later
+         */
+        if (iface_attr->dev_num_paths > max_dev_path) {
```

Comment:
why can't we mock more paths? What is the issue?

Reply:
There are 2 difficulties:

1. I noticed that dev_num_paths are implicitly "aligned" between IB transports (e.g. between `rc_mlx5` and `ud_mlx5`), in function `ucp_address_gather_devices`. There are 2 possible workarounds for this problem: either we mock both transports (which comes with just another problem: mocking of different TLs is hard to do); or we mock MD query to return some mock MD name for rc_mlx5.
I tried both options, I like more the second one, and by using it we can create as many dev paths as we want

2. wireup tries to connect all the lanes for all the paths, and it fails to do it with mocked paths (Destination is unreachable). It works well when real device supports that amount of paths. But when it exceeds it, then new paths are successfully created (e.g. mock_1:0/path2), but then wireup fails -> connection is terminating -> it hangs forever. Maybe it deserves a bit more investigation, I think we can find solution in the future. For now I just explicitly disable this

Reply:
can we force num paths to 1?

Reply:
We can force it in this PR if you want. But this is a serious limitation, which makes mocking useless for the task I'm working on. Having multiple paths is crucial to reproduce existing bugs in protocol selection:
- issue with lane sorting by BW, which happens only when device with multiple paths is present
- issue with incorrect performance estimation when multiple paths get assigned the entire device performance

Reply:
maybe you can set `IB_NUM_PATHS` to the value you need in the particular test?

Reply:
Thank you so much for this proposal, it works like a charm!
This is exactly what I needed to mock multiple paths. So I remove all these checks